### PR TITLE
Storage Volume resource migration to TPF

### PR DESCRIPTION
**Pull request by prajwalpatil25**
image

### CODE REVIEW

File: redfish/provider/resource_redfish_storage_volume.go [hunk 1]
```diff
@@ -1,5 +1,757 @@
...
+
+	var driveNames []string
+	diags.Append(d.Drives.ElementsAs(ctx, &driveNames, true)...)
+
+	volumeJobTimeout := d.ResetTimeout.ValueInt64()
+
+	// Get storage
+	storage, err := getStorage(service, storageID)
+	if err != nil {
+		diags.AddError("Error when retreiving the Systems from the Redfish API", err.Error())
```
Comment (rounak-adhikary) [inline redfish/provider/resource_redfish_storage_volume.go:506]:
```suggestion
		diags.AddError("Error retreiving storage details from the Redfish API", err.Error())
```

File: redfish/provider/resource_redfish_storage_volume.go [hunk 2]
```diff
@@ -1,5 +1,757 @@
...
+
+	// Get storage volumes
+	volumes, err := storage.Volumes()
+	if err != nil {
+		diags.AddError("there was an issue when retrieving volumes", err.Error())
+		return diags
+	}
+	volumeID, err := getVolumeID(volumes, volumeName)
+	if err != nil {
+		diags.AddError("Error. The volume ID with volume name %s on %s controller was not found", err.Error())
```
Comment (rounak-adhikary) [inline redfish/provider/resource_redfish_storage_volume.go:570]: These %s are not getting populated with info

File: redfish/provider/resource_redfish_storage_volume.go [hunk 3]
```diff
@@ -1,5 +1,757 @@
...
+func deleteRedfishStorageVolume(ctx context.Context, service *gofish.Service, d *models.RedfishStorageVolume) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	// Lock the mutex to avoid race conditions with other resources
+	redfishMutexKV.Lock(d.RedfishServer[0].Endpoint.ValueString())
+	defer redfishMutexKV.Unlock(d.RedfishServer[0].Endpoint.ValueString())
+
+	// Get vars from schema
+	applyTime := d.SettingsApplyTime.ValueString()
+	volumeJobTimeout := int64(d.VolumeJobTimeout.ValueInt64())
```
Comment (rounak-adhikary) [inline redfish/provider/resource_redfish_storage_volume.go:589]: No need for int64 conversion

File: redfish/provider/resource_redfish_storage_volume.go [hunk 4]
```diff
@@ -1,5 +1,757 @@
...
+	// Check if the volume exists
+	_, err := redfish.GetVolume(service.GetClient(), d.ID.ValueString())
+	if err != nil {
+		e, ok := err.(*redfishcommon.Error)
+		if !ok {
+			diags.AddError("There was an error with the API", err.Error())
+			return diags
+		}
+		if e.HTTPReturnedStatusCode == http.StatusNotFound {
+			log.Printf("Volume %s doesn't exist", d.ID.ValueString())
```
Comment (rounak-adhikary) [inline redfish/provider/resource_redfish_storage_volume.go:465]: tflog package please

File: redfish/provider/resource_redfish_storage_volume.go [hunk 5]
```diff
@@ -1,5 +1,757 @@
...
+			d.ID = types.StringValue("")
```
Comment (rounak-adhikary) [inline redfish/provider/resource_redfish_storage_volume.go:466]: I dont think this is doing the intended thing. I think we are supposed to recreate the volume if it has been somehow deleted.
One way to get it to work wd be to remove the state.

File: redfish/provider/resource_redfish_storage_volume.go (hunk 2 above)
Comment (prajwalpatil25) [reply redfish/provider/resource_redfish_storage_volume.go]: updated.

File: redfish/provider/resource_redfish_storage_volume.go (hunk 4 above)
Comment (prajwalpatil25) [reply redfish/provider/resource_redfish_storage_volume.go]: updated.

File: redfish/provider/resource_redfish_storage_volume.go (hunk 5 above)
Comment (prajwalpatil25) [reply redfish/provider/resource_redfish_storage_volume.go]: changed to remove resource if volume doesn't exist.

File: redfish/provider/resource_redfish_storage_volume.go (hunk 3 above)
Comment (prajwalpatil25) [reply redfish/provider/resource_redfish_storage_volume.go]: reverted

**Comment by prajwalpatil25** [comment]
```
=== RUN   TestAccRedfishStorageVolume_InvalidController
--- PASS: TestAccRedfishStorageVolume_InvalidController (2.65s)
=== RUN   TestAccRedfishStorageVolume_InvalidDrive
--- PASS: TestAccRedfishStorageVolume_InvalidDrive (3.50s)
=== RUN   TestAccRedfishStorageVolume_InvalidVolumeType
--- PASS: TestAccRedfishStorageVolume_InvalidVolumeType (4.08s)
=== RUN   TestAccRedfishStorageVolumeCreate_basic
--- PASS: TestAccRedfishStorageVolumeCreate_basic (183.94s)
=== RUN   TestAccRedfishStorageVolume_basic
--- PASS: TestAccRedfishStorageVolume_basic (194.35s)
=== RUN   TestAccRedfishStorageVolumeUpdate_basic
--- PASS: TestAccRedfishStorageVolumeUpdate_basic (297.40s)
=== RUN   TestAccRedfishStorageVolume_OnReset
--- PASS: TestAccRedfishStorageVolume_OnReset (684.04s)
PASS
        terraform-provider-redfish/redfish/provider     coverage: 19.5% of statements
ok      terraform-provider-redfish/redfish/provider     1369.999s       coverage: 19.5% of statements

```

**Review by Krishnan-Priyanshu — APPROVED** [review]

**Review by rounak-adhikary — APPROVED** [review]
lgtm

**Pull request merged**